compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirPrimaryConstructorSuperTypeChecker.kt

```diff
@@ -0,0 +1,77 @@
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.KtNodeTypes
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.checkers.toRegularClass
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.analysis.getParent
+import org.jetbrains.kotlin.fir.declarations.FirConstructor
+import org.jetbrains.kotlin.fir.declarations.FirRegularClass
+import org.jetbrains.kotlin.fir.declarations.isInterface
+import org.jetbrains.kotlin.fir.declarations.primaryConstructor
+import org.jetbrains.kotlin.fir.types.coneType
+import org.jetbrains.kotlin.fir.types.impl.FirImplicitAnyTypeRef
+import org.jetbrains.kotlin.utils.addToStdlib.lastIsInstanceOrNull
+
+object FirPrimaryConstructorSuperTypeChecker : FirRegularClassChecker() {
+    override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration.isInterface) {
+            return
+        }
+
+        val primaryConstructor = declaration.primaryConstructor
+
+        if (primaryConstructor == null) {
+            checkSupertypeInitializedWithoutPrimaryConstructor(declaration, reporter, context)
+        } else {
+            checkSuperTypeNotInitialized(primaryConstructor, declaration, context, reporter)
+        }
+    }
+
+    private fun checkSuperTypeNotInitialized(
+        primaryConstructor: FirConstructor,
+        regularClass: FirRegularClass,
+        context: CheckerContext,
+        reporter: DiagnosticReporter
+    ) {
+        val containingClass = context.containingDeclarations.lastIsInstanceOrNull<FirRegularClass>()
+        val delegatedConstructorCall = primaryConstructor.delegatedConstructor ?: return
+        // No need to check implicit call to the constructor of `kotlin.Any`.
+        if (delegatedConstructorCall.constructedTypeRef is FirImplicitAnyTypeRef) return
+        val superClass = delegatedConstructorCall.constructedTypeRef.coneType.toRegularClass(context.session) ?: return
+        // Subclassing a singleton should be reported as SINGLETON_IN_SUPERTYPE
+        if (superClass.classKind.isSingleton) return
+        if (regularClass.isEffectivelyExpect(containingClass, context) ||
+            regularClass.isEffectivelyExternal(containingClass, context)
+        ) {
+            return
+        }
+        if (delegatedConstructorCall.source?.elementType != KtNodeTypes.SUPER_TYPE_CALL_ENTRY) {
+            reporter.reportOn(
+                delegatedConstructorCall.constructedTypeRef.source?.getParent(KtNodeTypes.SUPER_TYPE_ENTRY, includingSelf = true),
```

Comment:
Why does `includingSelf` need to be `true` here? From what I can tell, it's always strictly the parent.

Reply:
That's right. I am only being permissive here. I guess the confusion introduced by this is not worth it. Let me remove that.